component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/reporter/AbstractMetricEventBuilder.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.am.analytics.publisher.reporter;
+
+import org.wso2.am.analytics.publisher.exception.MetricReportingException;
+
+import java.util.Map;
+
+/**
+ * Abstract implementation of {@link MetricEventBuilder}. Ensures that subclasses perform validations when adding
+ * elements.
+ */
+public abstract class AbstractMetricEventBuilder implements MetricEventBuilder {
+    @Override
+    public Map<String, Object> build() throws MetricReportingException {
+        if (validate()) {
+            return buildEvent();
+        }
+        throw new MetricReportingException("Validation failure occurred when building the event");
```

Comment:
is logging the exception handled elsewhere? If not shall we log at all places before throwing the exception?

Reply:
Figured that it should be handled by the code that calls this method. hence resolving.

Reply:
Log and throw in not a good practice. We can do that in this place though coz it is the outer most layer of the library. But I decided to transfer handling of that to the client. It doesn't provide us any edge by logging this